cmd/minishift/cmd/hostfolder/sftpd.go

```diff
@@ -159,47 +161,54 @@ func serveConnections(listener net.Listener, serverConfig *ssh.ServerConfig) {
 func serverConfig() *ssh.ServerConfig {
 	// An SSH server is represented by a ServerConfig, which holds certificate details and handles authentication
 	config := ssh.ServerConfig{
-		Config: ssh.Config{
-			MACs: []string{"hmac-sha1"},
-		},
 		PublicKeyCallback: keyAuth,
 	}
 
-	data, err := createPrivateKey()
-	if err != nil {
-		atexit.ExitWithMessage(1, fmt.Sprintf("Unable to create private key: %s", err))
-	}
-	hostPrivateKeySigner, err := ssh.ParsePrivateKey(data)
-	if err != nil {
-		atexit.ExitWithMessage(1, fmt.Sprintf("Unable to parse private key: %s", err))
+	if filehelper.Exists(minishiftConstants.ProfilePrivateKeyPath()) {
+		privateKeyBytes, err := ioutil.ReadFile(minishiftConstants.ProfilePrivateKeyPath())
+		if err != nil {
+			atexit.ExitWithMessage(1, fmt.Sprintf("Unable to read private key: %s", err))
+		}
+
+		privateKeySigner, err := ssh.ParsePrivateKey(privateKeyBytes)
+		if err != nil {
+			atexit.ExitWithMessage(1, fmt.Sprintf("Unable to parse private key: %s", err))
+		}
+		config.AddHostKey(privateKeySigner)
 	}
-	config.AddHostKey(hostPrivateKeySigner)
+
 	return &config
 }
 
-func keyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
-	// TODO Do proper key based authentication
-	// See also https://github.com/golang/crypto/blob/master/ssh/example_test.go
-	// As part of the key generation w/i the VM we create a public key which we need to store in a authorized_keys file
-	// for all profiles
-	permissions := &ssh.Permissions{
-		CriticalOptions: map[string]string{},
-		Extensions:      map[string]string{},
+// populateAuthorizedKeysMap populates the authorizedKeysMap map after reading the key from the
+// authorized_keys file.
+// authorized_keys file is maintained separately for each profile.
+func populateAuthorizedKeysMap() {
+	authorizedKeysBytes, err := ioutil.ReadFile(minishiftConstants.ProfileAuthorizedKeysPath())
+	if err != nil {
+		glog.Fatalf("Failed to load authorized_keys, err: %v", err)
 	}
-	return permissions, nil
-}
 
-func createPrivateKey() ([]byte, error) {
-	key, err := rsa.GenerateKey(rand.Reader, 2048)
-	if err != nil {
-		return nil, err
+	for len(authorizedKeysBytes) > 0 {
+		pubKey, _, _, rest, err := ssh.ParseAuthorizedKey(authorizedKeysBytes)
+		if err != nil {
+			glog.Fatal(err)
+		}
+
+		authorizedKeysMap[string(pubKey.Marshal())] = true
+		authorizedKeysBytes = rest
 	}
+}
 
-	pem := pem.EncodeToMemory(
-		&pem.Block{
-			Type:  "RSA PRIVATE KEY",
-			Bytes: x509.MarshalPKCS1PrivateKey(key),
-		},
-	)
-	return pem, nil
+func keyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
```

Comment:
What is the purpose of `conn ssh.Connmetadata` in this function, the formation of  `ssh.Permissions` does not seem to be dependent on it, only used for the error??

Reply:
Because, its a callback function and function signature is fixed
https://github.com/golang/crypto/blob/d6449816ce06963d9d136eee5a56fca5b0616e7e/ssh/server.go#L65-L67
```
// PasswordCallback, if non-nil, is called when a user
// attempts to authenticate using a password.
PasswordCallback func(conn ConnMetadata, password []byte) (*Permissions, error)
```

Reply:
Okay..